Remove unneeded require in plugin application.rb

skipkayhil opened the pull request:
This require was added in cfcea1d53ae5ce38a7cbeb41e05958dc009988b0, but at that time the generated Gemfile did not resolve using the gemspec. gemspec was added to the plugin's generated Gemfile later in a74e4736f95befa7a22c208019bf11a155ff7543. Since then, the Bundler.require in the plugin's generated config/application.rb will require the plugin and the extra require is unneeded.

The replaced test was added in adfa417fbbf6f670c3d10ed2c32c58bc1dc92c2e when the plugin generator's application.rb was replaced with the app generator's. Since the test is only concerned with the file content and not whether that content is useful, the test was replaced with an attempt to run the plugin's test, since this would verify that the plugin can be loaded within the dummy app.

The plugin helpers are added because the generated Gemfile is now used for resolution instead of Rails'. This makes the tests more accurate because real plugins will be resolving against their own Gemfiles. However, this also leads to a few issues solved by the new helpers:
- gemspecs are generated with TODOs that have to be fixed
- commands run in the plugin must have their Bundler environment reset because processes spawned in Rails' tests inherit BUNDLE_GEMFILE

Co-authored-by: Jonathan Hefner <jonathan@hefner.pro>

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* ~[ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.~

skipkayhil committed:
Remove unneeded require in plugin application.rb

This require was added in cfcea1d53ae5ce38a7cbeb41e05958dc009988b0, but
at that time the generated Gemfile did not resolve using the gemspec.
gemspec was added to the plugin's generated Gemfile later in
a74e4736f95befa7a22c208019bf11a155ff7543. Since then, the
Bundler.require in the plugin's generated config/application.rb will
require the plugin and the extra require is unneeded.

The replaced test was added in adfa417fbbf6f670c3d10ed2c32c58bc1dc92c2e
when the plugin generator's application.rb was replaced with the app
generator's. Since the test is only concerned with the file content and
not whether that content is useful, the test was replaced with an
attempt to run the plugin's test, since this would verify that the
plugin can be loaded within the dummy app.

The plugin helpers are added because the generated Gemfile is now used
for resolution instead of Rails'. This makes the tests more accurate
because real plugins will be resolving against their own Gemfiles.
However, this also leads to a few issues solved by the new helpers:
- gemspecs are generated with TODOs that have to be fixed
- commands run in the plugin must have their Bundler environment reset
because processes spawned in Rails' tests inherit BUNDLE_GEMFILE

Co-authored-by: Jonathan Hefner <jonathan@hefner.pro>

skipkayhil commented:
I'll look into the test failures, but I expect they are more examples of the generated plugins using Rails' Gemfile instead of the generated one.

skipkayhil commented:
cc @jonathanhefner I'd appreciate your opinion since this seems related to test infra you've worked on before. I _think_ the remaining failures could be fixed with something like #46478, but its all starting to get kind of complicated.

jonathanhefner commented:
> I _think_ the remaining failures could be fixed with something like #46478, but its all starting to get kind of complicated.

Yup, that was the missing piece!  But since most of these tests are not generator tests, we need to modify the `Gemfile` after the fact instead of stub `rails_gemfile_entry`.

I played around with the code a bit to verify.  I've pushed a new commit with changes.  Let me know what you think.  If it looks good, you or I can squash.

skipkayhil commented:
> I played around with the code a bit to verify. I've pushed a new commit with changes. Let me know what you think. If it looks good, you or I can squash.

It looks fantastic, thank you! I've squashed and updated the commit message with a description of the helpers

jonathanhefner commented:
Thank you, @skipkayhil! :electric_plug:
